README.md
# vue-lite-tooltip

> show tooltip to vue

[npm](https://www.npmjs.com/package/vue-lite-tooltip)

## EXAMPLE

https://s00d.github.io/vue-lite-tooltip/index.html

## Requirements

- vue: ^2.0.0

## Install

From npm:

``` sh
$ npm install vue-lite-tooltip --save
```

## use

``` js
import Vue from 'vue';
import VueLiteTooltip from 'vue-lite-tooltip';

Vue.use(VueLiteTooltip)

...

<button type="button" v-tooltip="{text: 'hello', position: 'right', width: 100}">test</button>

```


## Params

text - tooltip text
position - top, bottom, right, left
show - if false tooltip hidden
width - int size
color - text color
background - background color


## License

[MIT](https://opensource.org/licenses/MIT)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2019-09-16
### Removed
- old deps

## [1.1.0] - 2019-09-16
### Added
- ts









index.js
let tooltip = {};

let div = null;
let block = null;
let span = null;
let show = false;
let stop = false;

let showTooltip = function (opt) {
    if(stop) return div.style.opacity = 0;
    if(!show) return;
    setTimeout(() => {
        div.style.opacity = opt;
        opt = Math.round((opt+0.1)*10)/10;
        if(opt === 0.9) div.style.visibility = 'visible';
        if(opt <= 0.9) showTooltip(opt);
    }, 20)
};

let hideTooltip = function (opt) {
    if(stop) return div.style.opacity = 1;
    if(show) return;
    setTimeout(() => {
        div.style.opacity = opt;
        opt = Math.round((opt-0.1)*10)/10;
        if(opt === 0) {
            div.style.visibility = 'hidden';
            div.style.left = -100+'px';
            div.style.top = -100+'px';
        }
        if(opt >= 0) hideTooltip(opt);
    }, 20)
};

let genTooltip = function () {
    div = document.createElement("div");
    div.style.position = "absolute";
    div.classList.add("tooltip");
    div.style.visibility = 'hidden';
    div.style.left = -100+'px';
    div.style.top = -100+'px';
    div.style.width = 100+'px';
    div.style.opacity = 0.9;
    div.style.border = '1px solid #e4e4e4';

    div.style.zIndex = "1000";

    block = document.createElement("div");
    block.style.color =  'black';
    block.style.backgroundColor = "rgb(255, 255, 255)";
    block.style.padding = "7px 12px";
    block.style.maxWidth = "100%";
    block.classList.add("tooltip-inner");
    div.appendChild(block);

    span = document.createElement("span");

    block.appendChild(span);

    document.body.appendChild(div);

    div.onmouseover = function (event) {
        stop = true;
    };
    div.onmouseout = function (event) {
        stop = false;
        if(div !== null) {
            div.style.opacity = 1;
            hideTooltip(1);
        }
    };
};

tooltip.install = function(Vue) {
    genTooltip();
    Vue.directive('tooltip', function (el, binding) {
        el.onmouseover = function (event) {
            if ('show' in binding.value && !binding.value.show) return;
            span.innerHTML =  binding.value.text;
            if(!('position' in binding.value)) binding.value.position = 'top';
            if(!('width' in binding.value)) binding.value.width = '200';
            if('color' in binding.value) block.style.color = binding.value.color;
            if('background' in binding.value) block.style.backgroundColor = binding.value.color;
            div.style.width = binding.value.width+'px';
            switch (binding.value.position) {
                case 'top':
                    div.style.left = (event.pageX - (div.clientWidth / 2 - 5))+'px';
                    div.style.top = (event.pageY - (div.clientHeight + 10 ))+'px';
                    break;
                case 'bottom':
                    div.style.left = (event.pageX - (div.clientWidth / 2 - 5))+'px';
                    div.style.top = (event.pageY + 20)+'px';
                    break;
                case 'right':
                    div.style.left = (event.pageX + 30)+'px';
                    div.style.top = (event.pageY - (div.clientHeight / 2))+'px';
                    break;
                case 'left':
                    div.style.left = (event.pageX - (div.clientWidth + 10))+'px';
                    div.style.top = (event.pageY - (div.clientHeight / 2))+'px';
                    break;
                default:
                    break;
            }
            show = true;
            if(div !== null) {
                div.style.opacity = 0;
                showTooltip(0);
            }
        };
        el.onmouseout = function (event) {
            show = false;
            if(div !== null) {
                div.style.opacity = 1;
                hideTooltip(1);
            }
        };
        // el.onmousemove = function (event) {
        //     console.log(event);
        // };
    })
};

try {
    window.VueLiteTooltip = tooltip;
} catch(err) {}

export let VueLiteTooltip = tooltip;

src/index.ts
import Vue from 'vue'

class Tooltip {
    private div: HTMLElement|null = null;
    private block: HTMLElement|any = null;
    private span: HTMLElement|any = null;
    private show = false;
    private stop = false;
    // install(el: HTMLElement, node: HTMLElement) {
    private genTooltip() {
        this.div = document.createElement("div");
        this.div.style.position = "absolute";
        this.div.classList.add("tooltip");
        this.div.style.visibility = 'hidden';
        this.div.style.left = -100+'px';
        this.div.style.top = -100+'px';
        this.div.style.width = 100+'px';
        this.div.style.opacity = '0.9';
        this.div.style.border = '1px solid #e4e4e4';

        this.div.style.zIndex = "1000";

        this.block = document.createElement("div");
        this.block.style.color =  'black';
        this.block.style.backgroundColor = "rgb(255, 255, 255)";
        this.block.style.padding = "7px 12px";
        this.block.style.maxWidth = "100%";
        this.block.classList.add("tooltip-inner");
        this.div.appendChild(this.block);

        this.span = document.createElement("span");

        this.block.appendChild(this.span);

        document.body.appendChild(this.div);

        this.div.onmouseover = () => {
            this.stop = true;
        };
        this.div.onmouseout = () => {
            this.stop = false;
            if(this.div !== null) {
                this.div.style.opacity = '1';
                this.hideTooltip(1);
            }
        };
    }
    private showTooltip(opt: number): any {
        if(this.stop) return this.div!.style.opacity = '0';
        if(!this.show) return;
        setTimeout(() => {
            this.div!.style.opacity = opt.toString();
            opt = Math.round((opt+0.1)*10)/10;
            if(opt === 0.9) this.div!.style.visibility = 'visible';
            if(opt <= 0.9) this.showTooltip(opt);
        }, 20)
    }
    private hideTooltip(opt: number): any {
        if(this.stop) return this.div!.style.opacity = '1';
        if(this.show) return;
        setTimeout(() => {
            this.div!.style.opacity = opt.toString();
            opt = Math.round((opt-0.1)*10)/10;
            if(opt === 0) {
                this.div!.style.visibility = 'hidden';
                this.div!.style.left = -100+'px';
                this.div!.style.top = -100+'px';
            }
            if(opt >= 0) this.hideTooltip(opt);
        }, 20)
    }
    private readProps(value: {
        show?: boolean,
        position?: 'top'|'bottom'|'right'|'left',
        width?: string,
        color?:string,
        background?:string,
        title?:string,
        text:string,
    }) {
        if(value.show) value.show = true
        if(value.title) value.text = '<b>'+ value.title + '</b><br>' + value.text
        if(value.position) value.position = 'top'
        if(value.width) value.width = '200'
        return value;
    }
    install(_vue: typeof Vue) {
        this.genTooltip();
        _vue.directive('tooltip',  (el, binding) => {
            el.onmouseover = (event) => {
                const props = this.readProps(binding.value);
                if (!props.show) return;
                this.span.innerHTML = props.text;
                if(props.color) this.block.style.color = props.color;
                if(props.background) this.block.style.backgroundColor = props.background;
                this.div!.style.width = props.width+'px';
                switch (props.position) {
                    case 'top':
                        this.div!.style.left = (event.pageX - (this.div!.clientWidth / 2 - 5))+'px';
                        this.div!.style.top = (event.pageY - (this.div!.clientHeight + 10 ))+'px';
                        break;
                    case 'bottom':
                        this.div!.style.left = (event.pageX - (this.div!.clientWidth / 2 - 5))+'px';
                        this.div!.style.top = (event.pageY + 20)+'px';
                        break;
                    case 'right':
                        this.div!.style.left = (event.pageX + 30)+'px';
                        this.div!.style.top = (event.pageY - (this.div!.clientHeight / 2))+'px';
                        break;
                    case 'left':
                        this.div!.style.left = (event.pageX - (this.div!.clientWidth + 10))+'px';
                        this.div!.style.top = (event.pageY - (this.div!.clientHeight / 2))+'px';
                        break;
                    default:
                        break;
                }
                this.show = true;
                if(this.div !== null) {
                    this.div!.style.opacity = '0';
                    this.showTooltip(0);
                }
                if('click' in binding.value) {
                    el.onclick = el.onmouseout;
                }
            };

            if('click' in binding.value) {
                el.onclick = el.onmouseenter;
            }
            el.onmouseout = (event) => {
                this.show = false;
                if(this.div !== null) {
                    this.div.style.opacity = '1';
                    this.hideTooltip(1);
                }
            };
        })
    }
};

const tooltip = new Tooltip();

try {
    // @ts-ignore
    window.VueLiteTooltip = tooltip;
} catch(err) {
    console.log(err);
}

export const VueLiteTooltip = tooltip;
export default tooltip;
module.exports = tooltip;



index.build.js
(function (global, factory) {
    if (typeof define === "function" && define.amd) {
        define(['exports'], factory);
    } else if (typeof exports !== "undefined") {
        factory(exports);
    } else {
        var mod = {
            exports: {}
        };
        factory(mod.exports);
        global.index = mod.exports;
    }
})(this, function (exports) {
    'use strict';

    Object.defineProperty(exports, "__esModule", {
        value: true
    });
    var tooltip = {};

    var div = null;
    var block = null;
    var span = null;
    var show = false;
    var stop = false;

    var showTooltip = function showTooltip(opt) {
        if (stop) return div.style.opacity = 0;
        if (!show) return;
        setTimeout(function () {
            div.style.opacity = opt;
            opt = Math.round((opt + 0.1) * 10) / 10;
            if (opt === 0.9) div.style.visibility = 'visible';
            if (opt <= 0.9) showTooltip(opt);
        }, 20);
    };

    var hideTooltip = function hideTooltip(opt) {
        if (stop) return div.style.opacity = 1;
        if (show) return;
        setTimeout(function () {
            div.style.opacity = opt;
            opt = Math.round((opt - 0.1) * 10) / 10;
            if (opt === 0) {
                div.style.visibility = 'hidden';
                div.style.left = -100 + 'px';
                div.style.top = -100 + 'px';
            }
            if (opt >= 0) hideTooltip(opt);
        }, 20);
    };

    var genTooltip = function genTooltip() {
        div = document.createElement("div");
        div.style.position = "absolute";
        div.classList.add("tooltip");
        div.style.visibility = 'hidden';
        div.style.left = -100 + 'px';
        div.style.top = -100 + 'px';
        div.style.width = 100 + 'px';
        div.style.opacity = 0.9;
        div.style.border = '1px solid #e4e4e4';

        div.style.zIndex = "1000";

        block = document.createElement("div");
        block.style.color = 'black';
        block.style.backgroundColor = "rgb(255, 255, 255)";
        block.style.padding = "7px 12px";
        block.style.maxWidth = "100%";
        block.classList.add("tooltip-inner");
        div.appendChild(block);

        span = document.createElement("span");

        block.appendChild(span);

        document.body.appendChild(div);

        div.onmouseover = function (event) {
            stop = true;
        };
        div.onmouseout = function (event) {
            stop = false;
            if (div !== null) {
                div.style.opacity = 1;
                hideTooltip(1);
            }
        };
    };

    tooltip.install = function (Vue) {
        genTooltip();
        Vue.directive('tooltip', function (el, binding) {
            el.onmouseover = function (event) {
                if ('show' in binding.value && !binding.value.show) return;
                span.innerHTML = binding.value.text;
                if (!('position' in binding.value)) binding.value.position = 'top';
                if (!('width' in binding.value)) binding.value.width = '200';
                if ('color' in binding.value) block.style.color = binding.value.color;
                if ('background' in binding.value) block.style.backgroundColor = binding.value.color;
                div.style.width = binding.value.width + 'px';
                switch (binding.value.position) {
                    case 'top':
                        div.style.left = event.pageX - (div.clientWidth / 2 - 5) + 'px';
                        div.style.top = event.pageY - (div.clientHeight + 10) + 'px';
                        break;
                    case 'bottom':
                        div.style.left = event.pageX - (div.clientWidth / 2 - 5) + 'px';
                        div.style.top = event.pageY + 20 + 'px';
                        break;
                    case 'right':
                        div.style.left = event.pageX + 30 + 'px';
                        div.style.top = event.pageY - div.clientHeight / 2 + 'px';
                        break;
                    case 'left':
                        div.style.left = event.pageX - (div.clientWidth + 10) + 'px';
                        div.style.top = event.pageY - div.clientHeight / 2 + 'px';
                        break;
                    default:
                        break;
                }
                show = true;
                if (div !== null) {
                    div.style.opacity = 0;
                    showTooltip(0);
                }
            };
            el.onmouseout = function (event) {
                show = false;
                if (div !== null) {
                    div.style.opacity = 1;
                    hideTooltip(1);
                }
            };
            // el.onmousemove = function (event) {
            //     console.log(event);
            // };
        });
    };

    try {
        window.VueLiteTooltip = tooltip;
    } catch (err) {}

    var VueLiteTooltip = exports.VueLiteTooltip = tooltip;
});

